pkg/apis/configuration/v1/types.go

```diff
@@ -160,35 +160,6 @@ type KongPluginList struct {
 	Items []KongPlugin `json:"items"`
 }
 
-// Configuration contains a plugin configuration
-// +k8s:deepcopy-gen=false
-type Configuration map[string]interface{}
```

Comment:
Since we're only changing the representation (but the meaning of the type stays the same), I think it'd be better to maintain `type Configuration` and hide the conversion logic within it (by implementing a method):

```go
// Configuration is a representation of plugin configuration (an opaque JSON object).
type Configuration apiextensionsv1.JSON

// ToKongConfig converts raw Configuration into a configuration object.
func (c *Configuration) ToKongConfig(*kong.Configuration, error) {
	var result kong.Configuration
	if err := json.Unmarshal(c.Raw, &result); err != nil {
		return nil, err
	}
	return &res, nil
}
```

The alternative (replacing type `Configuration` by a raw opaque JSON object) loosens type safety and removes information about the type from `KongPlugin`.

Reply:
Can you elaborate on what issues you see the proposed change causing?

AFAIK (@hbagdi can maybe confirm) the only reason we had a separate type for these in the existing code was because [the deepcopy code generators cannot handle `map[string]interface{}`](https://github.com/kubernetes/gengo/issues/138). If they did, we'd have just used `map[string]interface{}` in `Kong(Cluster)Plugin` directly. We didn't have methods for `Configuration` other than the hand-written deepcopy methods needed to allow its use within the generated `Kong(Cluster)Plugin` deepcopy code.

I think a separate type with methods would be valuable here if we did something more complex with it than call `json.Unmarshal()` on it into a `go-kong.Configuration`, but we don't, so IMO the method would just add some unnecessary indirection.

Reply:
> what issues you see the proposed change causing?

Before this change, translation from the k8s api `Configuration` to `kong.Configuration` was a mere type conversion,
```go
plugin.Config = kong.Configuration(k8sPlugin.Config)
```

and the test for the config's presence was a comparison to nil:
```go
if k8sPlugin.Config != nil {
```

After this change, the former requires knowledge about the internal structure (that it is JSON and has the `Raw` field)
```go
err := json.Unmarshal(k8sPlugin.Config.Raw, &configuration)
if err != nil {
	return false, "could not unmarshal plugin configuration", err
}
plugin.Config = configuration
```

and the latter requires knowledge about the internal structure as well:
```go
if len(k8sPlugin.Config.Raw) > 0 {
```

This means that before this change, conversion and testing for presence were carried out by means of simple Go concepts that are universal across all types. After this change, conversion and testing for presence require the caller to know the inner structure of the object. Which is a code smell by itself, and I stand on the side of not introducing such issues into the codebase with incremental changes.

What my earlier comment overlooked is the test for presence.

So, to your point:
> I think a separate type with methods would be valuable here if we did something more complex

My PoV is that the caller's need to know the internal structure (the implementation of which is subject to change, as we see witt this PR) _is_ complex enough to justify abstracting it away.

Reply:
Refactored in 7ed00f3

Reply:
That commit actually breaks things. After it, `Raw` is still present, but is always empty.

That suggests whatever in API machinery populates that needs something more to recognize our type and stuff the JSON from the manifest into `Raw`, but I lack knowledge of what handles the conversion between manifest JSON to Go types in general and don't know what I should be looking for.

Trying to dig around in history to see what all was added for apiextensionsv1.JSON originally didn't yield anything of interest, as its creation predates the creation of apiextensions-apiserver and I'm not sure where it was created originally.

Reply:
@mflendrich I'm still unable to figure out why Raw doesn't get populated when this field is handled by a separate type.

Embedding apiextensionsv1.JSON directly within the Plugin types does work the way we expect, would you be okay with reverting 7ed00f3 and going with the original approach? I recognize that not exposing the internal structure is cleaner, but I don't know of a way to do that and still make the contents of field actually available.

Reply:
As discussed, the type not being aliased is good enough - the benefit of defining a separate type doesn't outweigh the investigation cost.